sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/QueueingBeamFnDataClient.java

```diff
@@ -98,35 +117,53 @@ private boolean allDone() {
    *
    * <p>All {@link InboundDataClient}s will be failed if processing throws an exception.
    *
-   * <p>This method is NOT thread safe. This should only be invoked by a single thread, and is
-   * intended for use with a newly constructed QueueingBeamFnDataClient in {@link
-   * ProcessBundleHandler#processBundle}.
+   * <p>This method is NOT thread safe. This should only be invoked once by a single thread.
    */
   public void drainAndBlock() throws Exception {
+    // If all of the inbound clients have already added to the queue, we don't
+    // use the poison to stop and instead poll. This avoids a possible deadlock
+    // if the queue is full and we are unable to queue the poison.
+    boolean requirePoison;
+    synchronized (inboundDataClients) {
+      Preconditions.checkState(!isDraining);
+      isDraining = true;
+      requirePoison = !inboundDataClients.isEmpty();
```

Comment:
question:

addPosion above is set based on `inboundDataClients.isEmpty() && isDraining`, and why requirePosion does not require the same check?

Reply:
there are two possibilities:
drainAndBlock is called while there are still clients -> last client is responsible for adding poison
drainAndBlock is called while there are no clients -> poison will not be added, drainAndBlock just drains existing queue

I took a stab at improving the comments throughout